src/MongoDB.Driver/ClientSessionHandle.cs

```diff
@@ -144,5 +244,153 @@ private TransactionOptions GetEffectiveTransactionOptions(TransactionOptions tra
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    CommitTransaction(cancellationToken);
+                    return false;
+                }
+                catch (Exception ex)
+                {
+                    if (ex is MongoException mongoException)
+                    {
+                        var now = _dateTimeProvider.Now;
+                        if (mongoException.HasErrorLabel(UnknownTransactionCommitResultLabel) &&
+                            !IsWriteConcernTimeoutError(mongoException) &&
+                            !IsTimeout(startTime, now))
+                        {
+                            continue;
+                        }
+
+                        if (mongoException.HasErrorLabel(TransientTransactionErrorLabel) && !IsTimeout(startTime, now))
+                        {
+                            // the transaction will be full processed once again.
+                            return true;
+                        }
+                    }
+
+                    throw;
+                }
+            }
+        }
+
+        private Task<bool> WithTransactionCommitProcessingWithRetryAsync(DateTime startTime, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    CommitTransactionAsync(cancellationToken).ConfigureAwait(false);
+                    return Task.FromResult(false);
+                }
+                catch (Exception ex)
+                {
+                    if (ex is MongoException mongoException)
+                    {
+                        var now = _dateTimeProvider.Now;
+                        if (mongoException.HasErrorLabel(UnknownTransactionCommitResultLabel) &&
+                            !IsWriteConcernTimeoutError(mongoException) &&
+                            !IsTimeout(startTime, now))
+                        {
+                            continue;
+                        }
+
+                        if (mongoException.HasErrorLabel(TransientTransactionErrorLabel) && !IsTimeout(startTime, now))
+                        {
+                            // the transaction will be full processed once again.
+                            return Task.FromResult(true);
+                        }
+                    }
+
+                    throw;
+                }
+            }
+        }
+
+        private bool IsTimeout(DateTime starTime, DateTime operationTime) => (operationTime - starTime) >= TimeSpan.FromSeconds(WithTransactionTimeout);
+
+        private bool IsWriteConcernTimeoutError(MongoException exception)
+        {
+            var writeConcernError = (exception as MongoWriteException)?.WriteConcernError ?? 
```

Comment:
> write concern timeout error (i.e. error code 64 and includes errInfo: { wtimeout: true })

I wasn't sure how this case would look like.
I found two exceptions which have `WriteConcern` field and check both of them here.
But It should be checked(somehow) once again to be sure that the server will throw exactly these errors.